lib/internal/url.js

```diff
@@ -286,38 +283,43 @@ Object.defineProperties(URL.prototype, {
         ret += '//';
         const has_username = typeof ctx.username === 'string';
         const has_password = typeof ctx.password === 'string' &&
-                             ctx.password !== '';
-        if (has_username || has_password) {
+                              ctx.password !== '';
+        if (options.auth && (has_username || has_password)) {
           if (has_username)
             ret += ctx.username;
           if (has_password)
             ret += `:${ctx.password}`;
           ret += '@';
         }
-        if (unicode) {
-          ret += punycode.toUnicode(this.hostname);
-          if (this.port !== undefined)
-            ret += `:${this.port}`;
-        } else {
-          ret += this.host;
-        }
+        ret += options.unicode ?
+          domainToUnicode(this.host) : this.host;
```

Comment:
Is the `unicode` option name from the spec, from compatibility with the old `url`, or is it up for debate? I don’t think it’s conveying the semantics very well (like, I have to look at the tests in this PR to see which behaviour maps to which value). How does something like `decodeIDNA: true/false` sound to you?